### PR TITLE
Upgrade to babel 7

### PR DESCRIPTION
**Pull request by cometkim**
#### Summary
Upgrade babel to 7 beta. (many other projects are already moved e.g. React Native)

See https://babeljs.io/docs/en/next/v7-migration

- babel core is peer dependency
- stage-x, es201x has been deprecated
- drop unused polyfills with `useBuiltIns: "usage"`
- much faster

Looks just fine for this package, but need to prepare for webapp and mobile either.

#### Ticket Link
None

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit tests passed
- [x] Ran `make flow` to ensure type checking passed
- ~~Added or updated unit tests (required for all new features)~~
- ~~All new/modified APIs include changes to the drivers~~

#### Test Information
This PR was tested on: Ubuntu Xenial

### CODE REVIEW

**Comment by jasonblais** [comment]
Hi @cometkim 

First, really appreciate your help with the refactoring and redux migration here: https://github.com/mattermost/mattermost-webapp/pull/1666

You have a great approach and are taking a comprehensive look at all the possible cases through your matrix 👍 

Second, we have recently discussed moving the webapp to use Babel 7. Part of this will be to determine if the ability to use native browser implementations for features and only polyfill when necessary will have significant performance improvements.

Given you've started the Babel 7 upgrade for redux, wondering if you were interested in participating in the effort to move the webapp to Babel 7? If so, I'd be happy to connect you with our team when you have time for it :)

**Comment by enahum** [comment]
BTW we just upgraded to RN 0.57.0 that uses babel 7

**Comment by cometkim** [comment]
Sounds great! Of course I interested to. actually already has tried to upgrade webapp project to use Babel 7 once, It has failed to build.

Anyway, I will resume this PR first (maybe this weekend..?). I think it wouldn't affect mobile and webapp project.

Thank you for notice @enahum, @jasonblais. please let me know if you started for webapp or mobile project :)

**Comment by jasonblais** [comment]
Thanks @cometkim! Sounds great. 

Given the mobile projects is now using react native v57, it contains the babel 7 upgrade. But once we've merged the Redux PR, we can work together on adding babel 7 for the webapp project.

**Comment by jasonblais** [comment]
@cometkim We're working on moving webapp to babel 7 now as well

**Comment by cometkim** [comment]
How is it going? It seems you've done at https://github.com/mattermost/mattermost-webapp/commit/a825d5c4976c4432cea899c3b4556b08d9f696a2.

**Comment by cometkim** [comment]
- Mocha / Chai / Nock
- ESLint
- Flow
- React
- reselect
- Webpack
- ws

are also need to be updated.

**Comment by jasonblais** [comment]
@cometkim Do those need to be updated on the webapp or redux repository?

**Comment by cometkim** [comment]
@jasonblais only on the redux repo and each can be handled in other PRs.

File: package.json [hunk 1]
```diff
@@ -24,17 +24,16 @@
     "shallow-equals": "1.0.0"
   },
   "devDependencies": {
-    "babel-cli": "6.26.0",
-    "babel-core": "6.26.3",
-    "babel-eslint": "8.2.3",
-    "babel-loader": "7.1.4",
+    "@babel/cli": "^7.1.2",
```
[the remark this answers is not in the document]
Comment (cometkim) [reply package.json]: @enahum yep.

**Comment by cometkim** [comment]
@enahum Any updates?

File: babel.config.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+module.exports = {
+    presets: [
```
[the remark this answers is not in the document]
Comment (enahum) [reply babel.config.js]: Just out of curiosity why it should match the webapp?

[the remark this answers is not in the document]
Comment (cometkim) [reply babel.config.js]: @crspeller I updated it to match with webapp's.

but I think that mattemrost-redux not only for mattermost-webapp and mattermost-mobile, but also it is a reference client implementation of the API. So future, We need to support a bit more compatibility and other specs like es module.

[the remark this answers is not in the document]
Comment (crspeller) [reply babel.config.js]: I was just asking about browser support. Which should probably be unified across our repositories. I agree that our support for this repository should be more broad.

**Pull request merged**